### PR TITLE
Default to script if script and family name match

### PR DESCRIPTION
**Pull request by bethanyaconnor**
As I was testing https://github.com/code-dot-org/code-dot-org/pull/41496, I noticed that if the family name of a script matches the script name, it can't find the script.

Our /s/ routes have some ambiguity. They can be used with the script name (`/s/coursea-2021`), the script id (`/s/1`), or the family name (`/s/coursea`). The first two options go through the same codepath and looks for a script who has a matching name or id to the parameter. However, the third option needs a different code path that would run first. This code path [only looks for stable scripts](https://github.com/code-dot-org/code-dot-org/blob/76286ba8af3d8e01f2248ba53ea1468937ed2513/dashboard/app/models/script.rb#L569). When the script is marked stable, it throws an error for calling `Script.get_with_cache` with a family name.

The solution here is to check for a script matching the name or id provided _before_ looking for scripts with a matching family name. This seems more in line with how these URLs are used and what users would expect in this case.

FWIW, this may seem like an edge case but for standalone scripts (like HOC scripts), the family name that makes the most sense is often the script name itself. 

## Testing story

tested locally, added a unit test

## Deployment strategy

As we currently don't have any scripts in this state, this should not affect any current behavior.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/script.rb [hunk 1]
```diff
@@ -515,16 +513,21 @@ def self.get_without_cache(id_or_name, with_associated_models: true)
...
+          # Populate cache on miss.
+          script_cache[cache_key] = get_without_cache(id_or_name)
+        end
+      else
+        get_without_cache(id_or_name, with_associated_models: false)
+      end
+    return script if script
+    if raise_exceptions
+      raise "Do not call Script.get_from_cache with a family_name. Call Script.get_unit_family_redirect_for_user instead.  Family: #{id_or_name}" if Script.family_names.include?(id_or_name)
+      raise ActiveRecord::RecordNotFound.new("Couldn't find Script with id|name=#{id_or_name}")
```
Comment (davidsbailey) [inline dashboard/app/models/script.rb:530]: this is getting a bit complicated.

Comment (davidsbailey) [reply dashboard/app/models/script.rb:530]: sorry, please ignore!

Comment (davidsbailey) [inline dashboard/app/models/script.rb:530]: apologies, I had drafted an incomplete thought here and clicked comment by accident. this makes this more complicated, but I don't see an obvious simpler way to handle this.

**Review by davidsbailey — APPROVED** [review]

File: dashboard/app/models/script.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/models/script.rb:530]: Yeah, ideally I wouldn't pass this parameter in but this method is called in a lot of places and I didn't want to risk breaking one of those places, at least in this PR.

**Pull request merged**